### ISSUE TITLE
retrieve clientmac/clientip variables with mode fas

### ISSUE THREAD

**Issue by electone77** [nodogsplash/nodogsplash#315]
hi,

I use nodogsplash with fas made and  its work.
I can retrieve $gatewayname,$authauction but i can't retrieve $clientip $clientmac
why ? :( 

Best regards.

**Comment by electone77** [comment]
sorry can you close this ticket.

**Comment by bluewavenet** [comment]
@electone77 
No problem. Ask here if you need any clarification.
For the benefit of others reading this, the documentation for using the NDS integrated FAS is here:
https://nodogsplashdocs.readthedocs.io/en/stable/fas.html

**Comment by dsmh** [comment]
Hi, I have the same problem, I need de clientemac passed to my FAS, is there any option to activate it?

**Comment by bluewavenet** [comment]
NDS forwards only clientip and gatewayname to your FAS.
You then pass clientip to ndsctl to obtain the token.
eg.
`ndsctl json $clientip | grep token | cut -c 10- | cut -c -8`

**Comment by bluewavenet** [comment]
So in a similar way to obtain the mac address:
`ndsctl json $clientip | grep mac | grep mac | cut -c 8- | cut -c -17`

**Comment by FenriLupus** [comment]
Hey, 
I have a same issue. 
I would like to work with the mac address of the client in a customized FAS. 

So please excuse the question, but I am not sure if I understood it correctly. 

The only way to access the client mac is to use: 
`ndsctl json $clientip | grep mac | cut -c 8- | cut -c -17`

So I need use samething like: 
`$clientmac = shell_exec("ndsctl json $clientip | grep mac | cut -c 8- | cut -c -17");`

Is that correct or is there an another workaround? 

HW: Raspberry 3B, SW: nodogsplash-master (V3.2.1)

**Comment by bluewavenet** [comment]
@FenriLupus 
There are many ways of obtaining the client mac address from the ip address.
For example you could get it from the arp tables or the dhcp database as well as asking ndsctl.
` ndsctl json $clientip | grep mac | cut -c 8- | cut -c -17` will work.
`ndsctl json $clientip | grep mac | awk -F '"' '{print $4}'` will work just as well.
`ip neigh | grep $clientip | awk -F ' ' '{print $5}'` will get it from arp.

On OpenWrt with Lighttpd and php7, the following example using ndsctl works:
`$clientmac = exec("sudo /usr/bin/ndsctl json $clientip | grep mac | cut -c 8- | cut -c -17");`
It would be similar on your RPi.
(Lighttpd runs as its own user (not root) so sudo is required).

**Comment by bluewavenet** [comment]
Just to be clear, this is not an issue with NDS/FAS. Parameters are deliberately not passed to FAS in the search string, for security reasons.

**Comment by FenriLupus** [comment]
@bluewavenet 
Thanks for your quick response.
`$clientmac= shell_exec(ndsctl json $clientip | grep mac | cut -c 8- | cut -c -17) `
Does not work on my device (probably a user config problem on my side). 
I spended now a while with troubleshooting, but I can't find the mistake.

But optaining the client mac address over arp works fine. 
`$clientmac = shell_exec(sudo ip neigh | grep $clientip | awk -F ' ' '{print $5}');`
So I will use this, thank you

**Comment by bluewavenet** [comment]
Good it works for you now :-D
For using the ndsctl method, try these two lines, assuming ndsctl is in /usr/bin on your system:

```
$cmd = "sudo /usr/bin/ndsctl json ".$clientip." | grep mac | cut -c 8- | cut -c -17";
$clientmac= exec($cmd);
```

**Comment by FenriLupus** [comment]
I was a little bit imprecise in my last contribution, sorry. 
I allready tryed to use:
`$clientmac = shell_exec("sudo /usr/bin/ndsctl json ".$clientip." | grep mac | cut -c 8- | cut -c -17");`
But I get still no output. 
The only way to get an ouput on my side is to use:
```
$output = shell_exec("sudo /usr/bin/ndsctl"); 
$output = "options:
  -s <path>           Path to the socket
  -h                  Print usage

commands:
  status              View the status of nodogsplash
  clients             Display machine-readable client list
  json [mac|ip|token] ..."
```


But if I add a command like `status`, `json` or `clients` I get no output from `shell_exec()`.

Anyway it works for me, it would be nice to know why I can't obtain ​​the client mac address over ndsctl but it is not necessary. And I have no idea anymore what I can try.

**Comment by bluewavenet** [comment]
You will need to get this working as you will have to obtain the client token from ndsctl to authenticate.

Does it work from the command line?
Have you tried exec() instead of shell_exec() ?

Try this as a diagnostic and see what the output is:
```
$cmd = "sudo /usr/bin/ndsctl json ".$clientip." | grep mac | cut -c 8- | cut -c -17";
$clientmac = exec($cmd, $output);
echo "clientmac=".$clientmac."!<br>";
print_r($output);
```

**Comment by FenriLupus** [comment]
Oh ^^
All commands are work fine in the shell. 
I tried both (`shell_exec();` and `exec();`).
I tried also same different versions of syntax and gave all users full permissions. 
Furthermore I troubleshoot PHP7 over `su www-data` but there are no logs available.

The FAS runs on an apache2 server that I host on the rapsberry.

The output of your diagnostic script is:` clientmac=!`

Does the `ndsctl` command need permissions to further resources (filse)?

**Comment by bluewavenet** [comment]
It is strange you do not get any output from:
`print_r($output);`
This implies something wrong with your php setup perhaps?
I have run this on apache and currently on lighttpd with no issues (on openwrt).

ndsctl normally runs as root.

See what you get with:
```
$cmd = "sudo /usr/bin/ndsctl json";
$clientmac = exec($cmd, $output);
echo "clientmac=".$clientmac."!<br>";
print_r($output);
```
Also look at the system logs - sudo should report what it is doing in there.

**Comment by FenriLupus** [comment]
Sorry for the late reply. 
At the moment the PHP and nodogsplash setup runs under default (exception of the few changes in the nds setup for FAS).

The output of your diagnostic script is:
```
clientmac=!
Array( )
```

I found the following in the system logs `/var/log/auth.log`: 
```
Mar 26 08:29:00 raspberrypi sudo: www-data : TTY=unknown ; PWD=/var/www/html/nodog ; USER=root ; COMMAND=/usr/bin/ndsctl json
Mar 26 08:29:00 raspberrypi sudo: pam_unix(sudo:session): session opened for user root by (uid=0)
Mar 26 08:29:00 raspberrypi sudo: pam_unix(sudo:session): session closed for user root
```
I am not so familiar with the syslog, but I would say it is normal?

**Comment by bluewavenet** [comment]
Ok then try this and see what it gives you, remembering to substitute a valid current client ip value in the first line:
```
$clientip="10.168.1.138";
$cmd = "sudo /usr/bin/ndsctl json ".$clientip." | grep mac | cut -c 8- | cut -c -17";
$clientmac = exec($cmd, $output);
echo "cmd=".$cmd."!<br>";
echo "clientmac=".$clientmac."!<br>";
print_r($output);

echo "<br><br>";
$cmd = "sudo /usr/bin/ndsctl json";
$clientmac = exec($cmd, $output);
echo "cmd=".$cmd."!<br>";
echo "clientmac=".$clientmac."!<br>";
print_r($output);
```
Here is the output I get:

```
cmd=sudo /usr/bin/ndsctl json 10.168.1.138 | grep mac | cut -c 8- | cut -c -17!
clientmac=74:de:2b:b1:70:33!
Array ( [0] => 74:de:2b:b1:70:33 )

cmd=sudo /usr/bin/ndsctl json!
clientmac=}!
Array ( [0] => 74:de:2b:b1:70:33 [1] => { [2] => "client_length": 2, [3] => "clients":{ [4] => "84:38:38:98:38:1a":{ [5] => "id":1, [6] => "ip":"10.168.1.90", [7] => "mac":"84:38:38:98:38:1a", [8] => "added":1553542261, [9] => "active":1553589683, [10] => "duration":47422, [11] => "token":"19cd83e6", [12] => "state":"Authenticated", [13] => "downloaded":543058, [14] => "avg_down_speed":91.61, [15] => "uploaded":22377, [16] => "avg_up_speed":3.78 [17] => }, [18] => "74:de:2b:b1:70:33":{ [19] => "id":23, [20] => "ip":"10.168.1.138", [21] => "mac":"74:de:2b:b1:70:33", [22] => "added":1553587005, [23] => "active":1553589683, [24] => "duration":2678, [25] => "token":"2570c14b", [26] => "state":"Authenticated", [27] => "downloaded":45398, [28] => "avg_down_speed":135.62, [29] => "uploaded":3034, [30] => "avg_up_speed":9.06 [31] => } [32] => } [33] => } ) 
```

**Comment by bluewavenet** [comment]
As you can see from my sample, exec() returns the last line of the output, but fills the array $output with the full result of the command, one line per element.

**Comment by FenriLupus** [comment]
I tryed this: 
```
$clientip="192.168.220.83";
$cmd = "sudo /usr/bin/ndsctl json ".$clientip." | grep mac | cut -c 8- | cut -c -17";
$clientmac = exec($cmd, $output);
echo "cmd=".$cmd."!<br>";
echo "clientmac=".$clientmac."!<br>";
print_r($output);
echo "<br><br>";

$cmd = "sudo /usr/bin/ndsctl json";
$clientmac = exec($cmd, $output);
echo "cmd=".$cmd."!<br>";
echo "clientmac=".$clientmac."!<br>";
print_r($output);
echo "<br><br>";

$cmd = "sudo /usr/bin/ndsctl";
exec($cmd, $output);
echo "cmd=".$cmd."!<br>";
print_r($output);
```
Output:
```
cmd=sudo /usr/bin/ndsctl json 192.168.220.83 | grep mac | cut -c 8- | cut -c -17!
clientmac=!
Array ( )

cmd=sudo /usr/bin/ndsctl json!
clientmac=!
Array ( ) 

cmd=sudo /usr/bin/ndsctl!
Array ( [0] => Usage: ndsctl [options] command [arguments] [1] => [2] => options: [3] => -s Path to the socket [4] => -h Print usage [5] => [6] => commands: [7] => status View the status of nodogsplash [8] => clients Display machine-readable client list [9] => json [mac|ip|token] Display client list in json format [10] => stop Stop the running nodogsplash [11] => auth mac|ip|token Authenticate user with specified mac, ip or token [12] => deauth mac|ip|token Deauthenticate user with specified mac, ip or token [13] => block mac Block the given MAC address [14] => unblock mac Unblock the given MAC address [15] => allow mac Allow the given MAC address [16] => unallow mac Unallow the given MAC address [17] => trust mac Trust the given MAC address [18] => untrust mac Untrust the given MAC address [19] => loglevel n Set logging level to n [20] => )
```

**Comment by bluewavenet** [comment]
My guess is this is caused by your sudo configuration.
To give a meaningful output, ndsctl must run as root.
Try modifying each $cmd to look like:
`$cmd = "sudo -u root /usr/bin/ndsctl json .....`

**Comment by mwarning** [comment]
What is the first line of your script (what shell do you use)?

**Comment by bluewavenet** [comment]
@mwarning 
It is a php script run (in this case) by php-cgi/apache and run as the apache user (not sure of the name).

I can replicate this exactly on openwrt with php7/lighttpd by forcing use of the lighttpd user (http).
ie:
`sudo -u http /usr/bin/ndsctl json ....`

**Comment by bluewavenet** [comment]
Running visudo and adding the line:
`root ALL=(ALL) ALL`
in the "User privilege specification" section will probably fix it....

**Comment by FenriLupus** [comment]
@bluewavenet If I run the commants with your modification (`$cmd = "sudo -u root /usr/bin/ndsctl json .....`) I get the same output as in my last contribution. 

I alreday tryed to edit visudo. 
At the moment it looks like that:   

```
# This file MUST be edited with the 'visudo' command as root.
#
# Please consider adding local content in /etc/sudoers.d/ instead of
# directly modifying this file.
#
# See the man page for details on how to write a sudoers file.
#
Defaults        env_reset
Defaults        mail_badpass
Defaults        secure_path="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:$

# Host alias specification

# User alias specification

# Cmnd alias specification
root ALL=(ALL) ALL


# User privilege specification
# Added by user (26.03.2019)
root    ALL=(ALL:ALL) ALL

# Allow members of group sudo to execute any command
%sudo   ALL=(ALL:ALL) ALL

# See sudoers(5) for more information on "#include" directives:

#includedir /etc/sudoers.d

# Added by user (22.03.2019)
www-data ALL=NOPASSWD: /usr/bin/ndsctl
www-data ALL=NOPASSWD: /usr/bin/ndsctl status
www-data ALL=NOPASSWD: /usr/bin/ndsctl json
www-data ALL=NOPASSWD: /usr/bin/nodogsplash
www-data ALL=NOPASSWD: ALL
root ALL=(ALL) ALL
```

**Comment by bluewavenet** [comment]
Take out everything you added to sudoers, then just add the single line:
`root ALL=(ALL) ALL`
in the "User privilege specification" section, then try again.

**Comment by FenriLupus** [comment]
visudo:
```
#
# This file MUST be edited with the 'visudo' command as root.
#
# Please consider adding local content in /etc/sudoers.d/ instead of
# directly modifying this file.
#
# See the man page for details on how to write a sudoers file.
#
Defaults        env_reset
Defaults        mail_badpass
Defaults        secure_path="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"

# Host alias specification

# User alias specification

# Cmnd alias specification
root ALL=(ALL) ALL


# User privilege specification
#Added by user (Florian, 26.03.2019)
root ALL=(ALL) ALL

# Allow members of group sudo to execute any command
%sudo   ALL=(ALL:ALL) ALL

# See sudoers(5) for more information on "#include" directives:

#includedir /etc/sudoers.d

# Added by user (Florian, 22.03.2019)
#www-data ALL=NOPASSWD: /usr/bin/ndsctl
#www-data ALL=NOPASSWD: /usr/bin/ndsctl status
#www-data ALL=NOPASSWD: /usr/bin/ndsctl json
#www-data ALL=NOPASSWD: /usr/bin/nodogsplash
#www-data ALL=NOPASSWD: ALL
#root ALL=(ALL) ALL

```

Output: 
```
cmd=sudo /usr/bin/ndsctl json 192.168.220.83 | grep mac | cut -c 8- | cut -c -17!
clientmac=!
Array ( )

cmd=sudo /usr/bin/ndsctl json!
clientmac=!
Array ( ) 

cmd=sudo /usr/bin/ndsctl!
Array ( )
```
Now even `sudo /usr/bin/ndsctl` is not working.

**Comment by bluewavenet** [comment]
> My guess is this is caused by your sudo configuration.
> To give a meaningful output, ndsctl must run as root.
> Try modifying each $cmd to look like:
> $cmd = "sudo -u root /usr/bin/ndsctl json .....

**Comment by FenriLupus** [comment]
@bluewavenet sorry I´ve forgot to edit the `Output `of my last contribution. 
I already tryed `$cmd = "sudo -u root /usr/bin/ndsctl json`.
Still no output.

**Comment by bluewavenet** [comment]
I have just tried this on a PCEngines apu2 board(x86-64 architecture) running opensuse and it works fine as it does on openwrt. 
Also people have reported this working on debian as well as raspbian.
You said it works on the command line so it can only be your sudo I'm afraid.
Can you start with a fresh operating system install?

**Comment by dokaga** [comment]
same problem,

i am having this error when i use shell_exec
ndsctl: nodogsplash probably not started (Error: No such file or directory)

**Comment by bluewavenet** [comment]
@dokaga 
Please start a new issue as this will most likely be related to your configuration and a new issue will be more useful to others.

**Comment by bawantha** [comment]
> ndsctl

what is ndsctl

**Comment by bluewavenet** [comment]
@bawantha 
ndsctl is the Nodogsplash command line utility. It is part of the package.
